src/backend/gporca/libgpopt/src/operators/CPhysicalHashJoin.cpp

```diff
@@ -373,8 +373,22 @@ CPhysicalHashJoin::PdshashedMatching(
 			}
 		}
 	}
+	// nulls colocated for inner hash joins, but not colocated in outer hash joins
+	BOOL fNullsColocated = true;
+
+	if (!m_is_null_aware &&
+		(COperator::EopPhysicalLeftOuterHashJoin == Eopid() ||
+		 COperator::EopPhysicalRightOuterHashJoin == Eopid()))
+	{
+		fNullsColocated = false;
+	}
+
 	// check if we failed to compute required distribution
-	if (pdrgpexpr->Size() != ulDlvrdSize)
+	// We could fail to find enough key expressions matching the source
+	// distribution, or we need the matching distribution have colocated nulls
+	// but input distribution's nulls are not colocated.
+	if (pdrgpexpr->Size() != ulDlvrdSize ||
+		(fNullsColocated && fNullsColocated != pdshashed->FNullsColocated()))
```

Comment:
How about simplifying here and below?
```suggestion
		(fNullsColocated && !pdshashed->FNullsColocated()))
```